README.md
# orite 
## A persistant rsync wrapper written in Python

Save a config file and then synchronise your remote and local folders and files. This project is an effort to centralise a sync approach, configure once, add options, and make it simple.

To do this using orite run:

    orite -^ 

To do this using rsync run:

    rsync --human-readable --info=flist --stats --archive --verbose --partial -ic --progress --dry-run /Users/username/Documents/websites/site_name/stack/folder/ username@127.68.551.54:/home/stack/folder --exclude-from="exclude.txt"


#### ōrite is a Māori word for ʻthe same’
It's pronounced [like this](https://s3.amazonaws.com/media.tewhanake.maori.nz/dictionary/4802.mp3) rather than ‘oh-right’.

***

## The problem

FTP apps like Transmit and Cyberduck are good apps and can synchronise content, but they are slow.

The CLI tool rsync is brilliant for this purpose in that it is lightning quick. The problem is that there is a lot to type in. There are a lot of flags to remember, as well as the username, IP address, the remote path, and the exclude file path. I also find it hard to remember to include a slash on the local and exclude a slash on the remote path. Meaning one can copy the entire repo into the wrong folder or just loose amongst your other stuff.

I used to do this kind of thing using Git and a Python library called fabric. Running the fabfile would commit, push, login to the remote server and then pull and restart. Whilst it was accurate it was a little clunky and again not super fast. I moved on from that to running a straight shell script. I ended up with variations of this script in each folder I made. 

This is where orite came in.

***

## Installation

    pip3 install orite

Or download the orite folder and alias orite.py in your bash profile.

    alias orite='python3 path-to-orite-folder/orite.py'
    
If you don’t have pip.

	sudo easy_install pip

You will also be required to have Python 3 installed.

## Update

	pip3 install orite --upgrade --no-cache-dir


## How to run?
Run orite in the directory above the folder that you would like to sync.

If you haven't used orite in this directory you will be prompted to lay down a config file and a exclude file.

To upload

    orite -^ 

To download

    orite -v

By default, orite will run in dry-run mode. Use the `-r` flag to override this and do a sync for real.

The help menu can be seen using

    orite -h


## Multiple servers – one config file

You can enable the config file to have more than one server or folder location. For example, my DEFAULT folder is my Django install and in a very separate location is my CSS folder. 

`orite -^` will upload (with --dry-run set) to my DEFAULT setting, and 

`orite CSS -^` will upload to my CSS section settings. 

Open the config file in your text editor. Add a section label at the bottom like so

`[CSS]`

Then add the `path_to_local_folder = /path/` on the next line and `path_to_remote_folder = /path/` to the line after. If your CSS settings are on the same server as your DEFAULT settings then you don't need to add those details again. Needless to say that if they are on a different server than add those settings to this section too.

Add as many sections as you like.

Run `orite -s` to get a print out of all the sections in your config file.


## Improvements/enhancements

Use orite with `watch` and have it watch a local directory for changes.

    watch -d 'ls path-to-dir/css/ | orite CSS -^'

If you are on a Mac and don't have watch in your CLI. Install it with [Homebrew](https://brew.sh/).

	brew install watch

For a brief overview of watch use `watch -h` and read something like [this](http://www.linfo.org/watch.html)

For more issues and suggested enhancements check out the issues tab.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## 0.0.17 - 2018-07-31
- Enabling multiple server sections in your config file

## 0.0.16 - 2018-07-28
- Fixing a issue with file name orite_exclude.txt

## 0.0.15 - 2018-07-24
- Fix a bug with copy_local

## 0.0.14 - 2018-07-24
### Added
- This CHANGELOG file.
- copy_local and sync_remote_copy methods and associated flag. To make an initial mirror of your remote folder in the folder orite__'folder-name'
- compare_local_to_remote_copy method to compare remote copy with your local folder.
- sftp method so that you can use your config file to run this command.

sync.py
#!/usr/local/bin/python3
import subprocess, os, sys

'''This will add the current working directory to the python path.
Meaning if this is run from the command line whatever directory the terminal is in.
'''
sys.path.insert(0, os.getcwd())


# Could do something like run the sync. If it has trouble connecting it could look for an RSA key.
# If it exists then copy it to the server.
# If it doesn't then make one and copy it to the server.


# Initialise: put down an initialising file to be filled out.
def initialise_settings():
	file = open('sync_settings.py', 'w')
	url = input( "Set your URL: " )
	file.write("url = '%s'\n" % url)
	# dev_url = input( "Set your Development URL: " )
	# file.write("dev_url = '%s'\n" % dev_url)
	file.write("dev_url = ''\n")
	current_ip_address = input("The current IP address: ")
	file.write("fromIP = '%s'\n" % current_ip_address)
	file.write("toIP = ''\n")
	file.write("docker_prefix = 'mdmediastack'\n")
	file.write("report_email = None\n")
	file.write("dropbox_access_token = None\n")

	# path_to_local_folder = input("The current IP address: ")
	path_to_local_folder = os.getcwd()
	file.write("path_to_local_folder = %s\n" % path_to_local_folder )
	
	path_to_remote_folder = "/"
	file.write("path_to_remote_folder = %s\n" % path_to_remote_folder )

	file.write("username = 'root'\n")
	file.write("server = fromIP\n")

	file.close()
	try: 
		exclude_file = open('_orite_exclude.txt')
	except FileNotFoundError:
		exclude_file = open('_orite_exclude.txt', 'w')
		exclude_file.write('''.git/
.DS_Store
.gitignore
# Cerificates
certs
dh.pem
key.pem
cert.pem
migrations
# Database files
*.sqlite
*.sqlite3
db/data
# Python cache files
*.pyc
__pycache__''')
		exclude_file.close()
	sys.stdout.write("The file 'sync_settings.py' and a default '_orite_exclude.txt' file have been made for you.")


try:
	from sync_settings import *
	settings_file_test = True
except ModuleNotFoundError:
	settings_file_test = False
	print('You have no sync_settings file.')
	# print('Initialise a settings file by running sync -i')
	initialise_settings()


# How does this directory differ from the remote one?


def local_to_remote(dry_run=True):
	print( 'Sync the local folder to the remote folder')
	vars = globals()
	print(vars['path_to_local_folder'])
	if dry_run:
		vars['dry_run'] = ' --dry-run'
	else:
		vars['dry_run'] = ''

	'''Add the dash to the end. The reason for this can be found on this page:
	https://www.digitalocean.com/community/tutorials/how-to-use-rsync-to-sync-local-and-remote-directories-on-a-vps
	'''
	if vars['path_to_local_folder'][-1] != '/':
		vars['path_to_local_folder'] = vars['path_to_local_folder'] + '/'

	'''Remove the dash from the end'''
	if vars['path_to_remote_folder'][-1] == '/':
		vars['path_to_remote_folder'] = vars['path_to_remote_folder'][:-1]

	'''Ensire that rsync is >= 3.1 for the --info=flist flag to work.
	Mac contains a version from 2006.
	This can be updated and documentation can be found here:
	https://f-a.nz/dev/update-macos-rsync-with-homebrew/
	'''
	command = 'rsync --human-readable --info=flist --stats --archive --verbose --partial -ic --progress{dry_run} {path_to_local_folder} {username}@{server}:{path_to_remote_folder} --exclude-from="sync_exclude.txt"'.format(**vars)
	# command = 'rsync --human-readable --info=flist --stats --archive --verbose --partial --progress{dry_run} {path_to_local_folder} {username}@{server}:{path_to_remote_folder} --exclude-from="sync_exclude.txt"'.format(**vars)
	# command = 'rsync --human-readable --stats --archive --verbose --partial --progress{dry_run} {path_to_local_folder} {username}@{server}:{path_to_remote_folder} --exclude-from="sync_exclude.txt"'.format(**vars)
	print( 'Running this command: ' + command )
	print('\n')
	return subprocess.call(command, shell=True)




def remote_to_local(dry_run=True):
	print( 'Sync the remote folder to the local folder')
	vars = globals()
	if dry_run:
		vars['dry_run'] = ' --dry-run'
	else:
		vars['dry_run'] = ''

	'''Remove the dash from the end'''
	if vars['path_to_local_folder'][-1] == '/':
		vars['path_to_local_folder'] = vars['path_to_local_folder'][:-1]
	'''Add the dash to the end'''
	if vars['path_to_remote_folder'][-1] != '/':
		vars['path_to_remote_folder'] = vars['path_to_remote_folder'] + '/'

	'''--info=flist flag only works in rysnc that is >= 3.1 Read comment in local_to_remote()'''
	command = 'rsync --human-readable --info=flist --stats --archive --verbose --partial --progress{dry_run} {username}@{server}:{path_to_remote_folder} {path_to_local_folder} --exclude-from="sync_exclude.txt"'.format(**vars)
	print( 'Running this command: ' + command )
	return subprocess.call(command, shell=True)


# rsync

# delete after flag


# sftp sync for servers that you can't get ssh access to.
def ssh_prompt():
	print('SSH command:')
	vars = globals()
	print('ssh %s@%s' % ( vars['username'], vars['server']) )



if __name__ == '__main__':

	import argparse
	'''
		Documentation here: https://docs.python.org/2/library/argparse.html
	'''
	parser = argparse.ArgumentParser(prog='Snyc', description='A wrapper for rsync with settings')
	parser.add_argument("-i", "--initialise_settings", help="Initialise the settings file")
	parser.add_argument("-v", "--from_remote_to_local", help="Sync the remote folder to the local folder", action="store_true")
	parser.add_argument("-^", "--from_local_to_remote", help="Sync the local folder to remote folder", action="store_true")
	parser.add_argument("-d", "--dry_run", help="Do a dry run. This is the default", action="store_true", default='')
	parser.add_argument("-r", "--for_real", help="Not a dry run, do it for real", action="store_true", default='')
	parser.add_argument("-ssh", "--ssh_prompt", help="Login using SSH", action="store_true", default='')
	args = parser.parse_args()


	if settings_file_test:
		dry_run = True
		if args.for_real:
			dry_run = False

		# remote_to_local(dry_run=True)
		if args.from_remote_to_local:
			remote_to_local(dry_run)
		elif args.from_local_to_remote:
			local_to_remote(dry_run)
		elif args.ssh_prompt:
			ssh_prompt()


# Make an invisible copy of the local folder to .<folder_name>--sync
# Sync from the remote to the copy
# run diff -r  <local-folder> .<local-folder>--sync/ --exclude-from <path-to>/sync_exclude.txt
# Bring this into sync and change so that it has more colour and is clearer what is what.

# Run diff but with numbers of the dirrences. Then the prompted can copy some of these accross.



orite/orite.py
#!/usr/local/bin/python3
'''
orite - is an opinionated rsync wrapper written in Python.

Whilst this is a Python module the intention is for it be run from the command line.
For a detailed write-up visit: https://github.com/meyouwe/orite
'''

from shutil import copyfile
import subprocess
import os
import sys
import argparse
import configparser


'''This will add the current working directory to the python path.
Meaning if this is run from the command line - whatever directory the terminal is in.
'''
sys.path.insert(0, os.getcwd())


exclude_file_name = "orite_exclude.txt"
config_file_name = "orite_config.ini"




def format_output(message, colour='34'):
	'''Makes the CLI output bold and gives it a colour.
	The default colour - 34 is blue.
	Colour numbers can be found here: https://en.wikipedia.org/wiki/ANSI_escape_code'''
	return "\x1b[1;" + colour + "m" + message + "\x1b[0m"




class initialise():
	'''This class has methods that check to see if the config file and exclude file is in the path where the script is running from. If they aren't then there are two methods that will initialise these files.'''
	def __init__(self):
		self.config_file_name = config_file_name
		self.exclude_file_name = exclude_file_name


	def initialise_config_file(self):
		''' Prompt the user and lay down and configuration file.'''

		config = configparser.ConfigParser()
		c = {}
		c['url'] = input( "Set your URL: " )

		c['remote_server'] = input("Set the remote IP address: ")

		c['username'] = input("Set your username: ")

		'''pwd and getcwd() doesn't escape spaces and rsync requires spaces to be escaped. This escapes spaces.''' 
		default_path_to_local_folder = '\ '.join(os.getcwd().split(' '))
		continue_question = input("Is your local folder: %s  [y/n]" % default_path_to_local_folder)
		if continue_question.lower() == 'n':
			c['path_to_local_folder'] = input("Set your local folder (escape spaces with a \): ")
		else:
			c['path_to_local_folder'] = default_path_to_local_folder
		
		c['path_to_remote_folder'] = input("Set your remote folder path: ")

		config['DEFAULT'] = c
		with open(config_file_name, 'w') as configfile:
			config.write(configfile)

		sys.stdout.write(
			format_output("The orite configuration file '{config_file_name}' has been made. Edit it how you like.\n".format(
				**globals()
			)
		))


	def config_file_exists(self):
		"""Test to see if the config file exists. If id doesn't make one."""
		try:
			config_file_test = open(self.config_file_name)
			config_file_test.close()
		except FileNotFoundError:
			continue_question = input(format_output('You have no config file in this directory. Would you like to make one? [y/n] \n'))
			if continue_question.lower() == 'n':
				sys.exit()
			else:
				self.initialise_config_file()



	def config_file_sections(self):
		"""Return the names of each section."""
		config_file_test = open(self.config_file_name, 'r')
		lines = config_file_test.readlines()
		config_file_test.close()
		for line in lines:
			if line[0] == '[':
				sys.stdout.write(line)



	def copy_exclude_file(self):
		'''Copy across the default exclude file as a starting point.
		I have seen .gitignore used as an exclude file. This is an interesting idea but not there will be differences between these two.
		'''
		script_dest = os.path.dirname(os.path.realpath(__file__))
		copyfile(os.path.join(script_dest, self.exclude_file_name), os.path.join(os.getcwd(), self.exclude_file_name))
		sys.stdout.write(format_output("A default exclude file has been made at '%s'. Edit it how you like.\n" % self.exclude_file_name))


	def exclude_file_exists(self):
		'''Test to see if the exclude file exists. If not copy the default file across.'''
		try:
			exclude_file_test = open(self.exclude_file_name)
			exclude_file_test.close()
		except FileNotFoundError:
			self.copy_exclude_file()




class commands():

	def __init__(self, username, remote_server, local_path, remote_path, dry_run):
		self.username = username
		self.remote_server = remote_server
		self.local_path = local_path
		self.remote_path = remote_path
		self.exclude_file_name = exclude_file_name

		if dry_run:
			self.dry_run = ' --dry-run'
		else:
			self.dry_run = ''

		if self.local_path[-1] == '/':
			local_path = self.local_path[:-1]
		else:
			local_path = self.local_path
		self.local_folder = os.path.basename(local_path)
		self.local_folder_copy_name = 'orite__%s' % self.local_folder


	def local_to_remote(self):
		'''Sync from the local path to the remote server'''
		sys.stdout.write('\nSync the local folder to the remote folder\n')

		'''Add the dash to the end. The reason for this can be found on this page: https://www.digitalocean.com/community/tutorials/how-to-use-rsync-to-sync-local-and-remote-directories-on-a-vps'''
		if self.local_path[-1] != '/':
			self.local_path = self.local_path + '/'
		'''Remove the dash from the end'''
		if self.remote_path[-1] == '/':
			self.remote_path = self.remote_path[:-1]

		'''Ensure that rsync is >= 3.1 for the --info=flist flag to work.
		Mac contains a version from 2006.
		This can be updated and documentation can be found here: https://f-a.nz/dev/update-macos-rsync-with-homebrew/
		'''
		command = 'rsync --human-readable --info=flist --stats --archive --verbose --partial -ic --progress{o.dry_run} {o.local_path} {o.username}@{o.remote_server}:{o.remote_path} --exclude-from="{o.exclude_file_name}"'.format(o=self)
		# Option for rsync <3.1
		# command = 'rsync --human-readable --stats --archive --verbose --partial --progress{o.dry_run} {o.local_path} {o.username}@{o.remote_server}:{o.remote_path} --exclude-from="{o.exclude_file_name}"'.format(o=self)
		sys.stdout.write(format_output('\nRunning this command: \n') + command + '\n\n')
		return subprocess.call(command, shell=True)


	def remote_to_local(self, local_path=False):
		'''Sync from the remote server to the local directory'''
		sys.stdout.write('\nSync the remote folder to the local folder\n')

		if not local_path:
			local_path = self.local_path

		'''Remove the dash from the end'''
		if local_path[-1] == '/':
			local_path = local_path[:-1]
		'''Add the dash to the end'''
		if self.remote_path[-1] != '/':
			self.remote_path = self.remote_path + '/'

		'''--info=flist flag only works in rysnc that is >= 3.1 Read comment in local_to_remote()'''
		command = 'rsync --human-readable --info=flist --stats --archive --verbose --partial --progress{o.dry_run} {o.username}@{o.remote_server}:{o.remote_path} {local_path} --exclude-from="{o.exclude_file_name}"'.format(o=self, **locals())
		sys.stdout.write(format_output('\nRunning this command: \n') + command + '\n\n')
		return subprocess.call(command, shell=True)


	def copy_local(self):
		if not os.path.isdir(self.local_folder_copy_name):
			command = 'cp -r {o.local_path} {o.local_folder_copy_name}'.format(o=self)
			subprocess.call(command, shell=True)


	def sync_remote_copy(self):
		local_copy_folder_path = '\ '.join(os.getcwd().split(' '))
		if local_copy_folder_path[-1] == '/':
			local_copy_folder_path = local_copy_folder_path + self.local_folder_copy_name
		else:
			local_copy_folder_path = local_copy_folder_path + '/' + self.local_folder_copy_name
		self.remote_to_local(local_path=local_copy_folder_path)


	def compare_local_to_remote_copy(self):
		'''Compare the local folder to the local remote copy.'''
		command = 'diff --recursive --new-file  {o.local_path} {o.local_folder_copy_name}/ --exclude-from {o.exclude_file_name}'.format(o=self, **locals())
		sys.stdout.write(format_output('\nRunning this command: \n') + command + '\n\n')
		return subprocess.call(command, shell=True)


	def ssh_into_remote(self):
		'''ssh into the exact path on the remote server using your config file.'''
		sys.stdout.write(format_output('Running this command: \n'))
		command = 'ssh {o.username}@{o.remote_server} -t "cd {o.remote_path}; bash --login"'.format(o=self)
		sys.stdout.write(command + '\n')
		return subprocess.call(command, shell=True)


	def sftp_into_remote(self):
		'''sftp into the exact path on the remote server using your config file.'''
		sys.stdout.write(format_output('Running this command: \n'))
		command = 'sftp {o.username}@{o.remote_server}:{o.remote_path}'.format(o=self)
		sys.stdout.write(command + '\n')
		return subprocess.call(command, shell=True)




def main():
	'''FYI argparse documentation is here: https://docs.python.org/3.6/library/argparse.html
	Note that nargs='?' = 0 or 1 and nargs='*' is 0 or more arguments'''
	parser = argparse.ArgumentParser(prog='orite', description='A Python wrapper mainly around rsync with configuration files.')
	parser.add_argument("server", metavar='server', help="Config file section/server name", default='DEFAULT', nargs='?') 
	parser.add_argument("-s", "--sections", help="Print the section of the config file to the screen", action="store_true")
	parser.add_argument("-v", "--remote_to_local", help="Sync the remote folder to the local folder", action="store_true")
	parser.add_argument("-^", "--local_to_remote", help="Sync the local folder to the remote folder", action="store_true")
	parser.add_argument("-d", "--dry_run", help="Do a dry run. This is the default", action="store_true", default='')
	parser.add_argument("-r", "--for_real", help="Not a dry run, do it for real", action="store_true", default='')
	parser.add_argument("--ssh", help="Login using SSH", action="store_true", default='')
	parser.add_argument("--sftp", help="Login using SFTP", action="store_true", default='')

	parser.add_argument("-C", "--initial_copy", help="Copy the local folder for use in diff.", action="store_true", default='')
	parser.add_argument('-R', "--remote_to_remote_copy", help="Sync the remote folder to the local remote copy", action="store_true", default='')
	parser.add_argument('-D', "--diff", help="Compare the local folder to the local remote copy", action="store_true", default='')
	args = parser.parse_args()

	init = initialise()
	init.config_file_exists()
	init.exclude_file_exists()

	if args.sections:
		init.config_file_sections()
	else:
		if args.for_real:
			dry_run = False
		else:
			dry_run = True

		config = configparser.ConfigParser()
		config.read(config_file_name)

		config_section = args.server
		try:
			config = config[config_section]
			username = config['username']
			remote_server = config['remote_server']
			local_path = config['path_to_local_folder']
			remote_path = config['path_to_remote_folder']
			
			com = commands(username, remote_server, local_path, remote_path, dry_run)

		except KeyError:
			sys.stdout.write(format_output('A keyword is missing in your config file\n', '31'))
			sys.stdout.write('Check the spelling or remove the config file altogther and re-initialise orite.\n')
			sys.exit()


		if args.remote_to_local:
			com.remote_to_local()
		elif args.local_to_remote:
			com.local_to_remote()
		elif args.ssh:
			com.ssh_into_remote()
		elif args.sftp:
			com.sftp_into_remote()
		elif args.initial_copy:
			com.copy_local()
		elif args.remote_to_remote_copy:
			com.sync_remote_copy()
		elif args.diff:
			com.compare_local_to_remote_copy()
		else:
			sys.stdout.write(format_output('Use orite -h to see which options are available\n'))



if __name__ == '__main__':
	main()